File: capacitor/tessera-mrz/android/src/main/kotlin/io/lightine/tessera/capacitor/mrz/TesseraMrzPlugin.kt
package io.lightine.tessera.capacitor.mrz

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.activity.result.ActivityResult
import com.getcapacitor.JSObject
import com.getcapacitor.PermissionState
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback
import com.getcapacitor.annotation.PluginMethod

@CapacitorPlugin(
    name = "TesseraMrz",
    permissions = [Permission(strings = [Manifest.permission.CAMERA], alias = CAMERA_PERMISSION)],
)
class TesseraMrzPlugin : Plugin() {
    private var activeScanCall: PluginCall? = null
    private var cancelRequested: Boolean = false

    @PluginMethod
    fun scan(call: PluginCall) {
        if (activeScanCall != null) {
            call.reject("A scanner session is already active.", "SCANNER_ALREADY_ACTIVE")
            return
        }

        val request = parseRequest(call) ?: return
        TesseraScannerActivity.clearPendingCancellation()
        activeScanCall = call
        cancelRequested = false

        if (getPermissionState(CAMERA_PERMISSION) == PermissionState.GRANTED) {
            launchScanner(call, request)
        } else {
            requestPermissionForAlias(CAMERA_PERMISSION, call, "cameraPermissionResult")
        }
    }

    @PluginMethod
    fun cancelScan(call: PluginCall) {
        val scanCall = activeScanCall
        if (scanCall != null && !TesseraScannerActivity.cancelActiveOrNext()) {
            // Android's permission dialog belongs to the OS and cannot be dismissed by the plugin.
            // Resolve the pending scan as cancelled as soon as that dialog returns.
            cancelRequested = true
        }
        call.resolve()
    }

    @PluginMethod
    fun isSupported(call: PluginCall) {
        val cameraAvailable = context.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)
        val supported = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && cameraAvailable
        val result = JSObject().put("supported", supported)
        if (!supported) {
            result.put(
                "reason",
                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
                    "unsupported-os-version"
                } else {
                    "camera-unavailable"
                },
            )
        }
        call.resolve(result)
    }

    @PluginMethod
    fun getPluginInfo(call: PluginCall) {
        call.resolve(
            JSObject()
                .put("name", "capacitor-tessera-mrz")
                .put("version", "0.1.0")
                .put("tesseraVersion", "0.5.0")
                .put("capacitorMajor", 8)
                .put("platform", "android"),
        )
    }

    @PermissionCallback
    private fun cameraPermissionResult(call: PluginCall) {
        if (activeScanCall !== call) return
        if (cancelRequested) {
            call.resolve(JSObject().put("status", "cancelled"))
            clearActiveScan()
            return
        }
        if (getPermissionState(CAMERA_PERMISSION) != PermissionState.GRANTED) {
            call.reject("Camera permission was denied.", "CAMERA_PERMISSION_DENIED")
            clearActiveScan()
            return
        }
        val request = parseRequest(call)
        if (request == null) {
            clearActiveScan()
            return
        }
        launchScanner(call, request)
    }

    @ActivityCallback
    private fun scannerResult(
        call: PluginCall?,
        result: ActivityResult,
    ) {
        val scanCall = call ?: activeScanCall
        clearActiveScan()
        if (scanCall == null) return

        val data = result.data
        if (result.resultCode == Activity.RESULT_OK) {
            val json = data?.getStringExtra(TesseraScannerActivity.EXTRA_RESULT_JSON)
            if (json == null) {
                scanCall.reject("The scanner returned no result.", "INTERNAL_ERROR")
            } else {
                runCatching { JSObject(json) }
                    .onSuccess(scanCall::resolve)
                    .onFailure { scanCall.reject("The scanner returned an invalid result.", "INTERNAL_ERROR") }
            }
            return
        }

        val code = data?.getStringExtra(TesseraScannerActivity.EXTRA_ERROR_CODE) ?: "INTERNAL_ERROR"
        scanCall.reject(errorMessage(code), code)
    }

    private fun launchScanner(
        call: PluginCall,
        request: ScanRequest,
    ) {
        val intent = Intent(context, TesseraScannerActivity::class.java)
        request.writeTo(intent)
        startActivityForResult(call, intent, "scannerResult")
    }

    private fun parseRequest(call: PluginCall): ScanRequest? {
        val documentTypesArray = call.getArray("documentTypes")
        val documentTypes =
            try {
                val rawValues = documentTypesArray?.toList<Any?>().orEmpty()
                if (rawValues.any { it !is String }) {
                    throw IllegalArgumentException("documentTypes contains a non-string value")
                }
                rawValues.map { it as String }.toSet()
            } catch (_: Exception) {
                call.reject("documentTypes must be an array of strings.", "UNSUPPORTED_OPTION")
                return null
            }
        if (documentTypesArray != null && documentTypes.isEmpty()) {
            call.reject("documentTypes must not be empty.", "UNSUPPORTED_OPTION")
            return null
        }
        if (!documentTypes.all { it in SUPPORTED_DOCUMENT_TYPES }) {
            call.reject("documentTypes contains an unsupported value.", "UNSUPPORTED_OPTION")
            return null
        }

        val camera = call.getString("camera", "back")
        if (camera != "back") {
            call.reject("Only the back camera is supported by Tessera 0.5.0.", "UNSUPPORTED_OPTION")
            return null
        }

        val timeoutMs = call.getLong("timeoutMs")
        if (timeoutMs != null && timeoutMs <= 0L) {
            call.reject("timeoutMs must be greater than zero.", "UNSUPPORTED_OPTION")
            return null
        }

        return ScanRequest(
            documentTypes = documentTypes,
            allowPartialResults = call.getBoolean("allowPartialResults", false) ?: false,
            timeoutMs = timeoutMs,
            torchEnabled = call.getBoolean("torchEnabled", false) ?: false,
        )
    }

    private fun clearActiveScan() {
        activeScanCall = null
        cancelRequested = false
    }

    private fun errorMessage(code: String): String =
        when (code) {
            "SCAN_TIMEOUT" -> "The scanner session timed out."
            "CAMERA_UNAVAILABLE" -> "The camera is unavailable."
            "CAMERA_PERMISSION_DENIED" -> "Camera permission was denied."
            else -> "The scanner could not complete."
        }

    private companion object {
        const val CAMERA_PERMISSION = "camera"
        val SUPPORTED_DOCUMENT_TYPES = setOf("passport", "id-card", "visa")
    }
}

File: capacitor/tessera-mrz/ios/Sources/TesseraMrzPlugin/TesseraMrzPlugin.swift
import AVFoundation
import Capacitor
import Foundation
import Tessera
import TesseraUI

@objc(TesseraMrzPlugin)
public class TesseraMrzPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "TesseraMrzPlugin"
    public let jsName = "TesseraMrz"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "scan", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelScan", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isSupported", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginInfo", returnType: CAPPluginReturnPromise),
    ]

    private var activeCall: CAPPluginCall?
    private var scannerController: TesseraScannerController?
    private var cancelRequested = false

    @objc func scan(_ call: CAPPluginCall) {
        guard let request = ScanRequest.parse(call) else { return }
        DispatchQueue.main.async { [weak self] in
            self?.beginScan(call, request: request)
        }
    }

    @objc func cancelScan(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self else {
                call.resolve()
                return
            }
            if let scannerController = self.scannerController {
                scannerController.cancel()
            } else if self.activeCall != nil {
                // The system permission alert cannot be dismissed programmatically. Resolve the scan as
                // cancelled as soon as the alert returns.
                self.cancelRequested = true
            }
            call.resolve()
        }
    }

    @objc func isSupported(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            let supported = AVCaptureDevice.default(for: .video) != nil
            var result: JSObject = ["supported": supported]
            if !supported {
                result["reason"] = "camera-unavailable"
            }
            call.resolve(result)
        }
    }

    @objc func getPluginInfo(_ call: CAPPluginCall) {
        call.resolve([
            "name": "capacitor-tessera-mrz",
            "version": "0.1.0",
            "tesseraVersion": "0.5.0",
            "capacitorMajor": 8,
            "platform": "ios",
        ])
    }

    @MainActor
    private func beginScan(_ call: CAPPluginCall, request: ScanRequest) {
        guard activeCall == nil else {
            call.reject("A scanner session is already active.", "SCANNER_ALREADY_ACTIVE")
            return
        }
        activeCall = call
        cancelRequested = false

        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            presentScanner(request)
        case .notDetermined:
            let callbackId = call.callbackId
            AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
                DispatchQueue.main.async {
                    self?.permissionRequestCompleted(callbackId: callbackId, granted: granted, request: request)
                }
            }
        case .denied, .restricted:
            rejectActive("Camera permission was denied.", code: "CAMERA_PERMISSION_DENIED")
        @unknown default:
            rejectActive("Camera access is unavailable.", code: "CAMERA_UNAVAILABLE")
        }
    }

    @MainActor
    private func permissionRequestCompleted(callbackId: String, granted: Bool, request: ScanRequest) {
        guard activeCall?.callbackId == callbackId else { return }
        if cancelRequested {
            activeCall?.resolve(["status": "cancelled"])
            clearActiveScan()
        } else if granted {
            presentScanner(request)
        } else {
            rejectActive("Camera permission was denied.", code: "CAMERA_PERMISSION_DENIED")
        }
    }

    @MainActor
    private func presentScanner(_ request: ScanRequest) {
        guard let presenter = bridge?.viewController else {
            rejectActive("The scanner could not be presented.", code: "SCANNER_INITIALIZATION_FAILED")
            return
        }

        let controller = TesseraScannerController(request: request) { [weak self] result in
            self?.scannerCompleted(result)
        }
        guard controller.present(from: presenter) else {
            rejectActive("The scanner could not be presented.", code: "SCANNER_INITIALIZATION_FAILED")
            return
        }
        scannerController = controller
    }

    @MainActor
    private func scannerCompleted(_ result: TesseraUIResult) {
        guard let call = activeCall else { return }
        switch result {
        case let .confirmed(decoded):
            guard let mapped = MrzResultMapper.map(decoded.parse) else {
                call.reject("The MRZ could not be parsed.", "MRZ_PARSE_FAILED")
                clearActiveScan()
                return
            }
            call.resolve(mapped.toJSObject())
        case let .cancelled(reason):
            switch reason {
            case .userDismissed:
                call.resolve(["status": "cancelled"])
            case .timedOut:
                call.reject("The scanner session timed out.", "SCAN_TIMEOUT")
            case .cameraUnavailable:
                call.reject("The camera is unavailable.", "CAMERA_UNAVAILABLE")
            case .permissionDenied:
                call.reject("Camera permission was denied.", "CAMERA_PERMISSION_DENIED")
            }
        }
        clearActiveScan()
    }

    @MainActor
    private func rejectActive(_ message: String, code: String) {
        activeCall?.reject(message, code)
        clearActiveScan()
    }

    @MainActor
    private func clearActiveScan() {
        activeCall = nil
        scannerController = nil
        cancelRequested = false
    }
}
